lib/at_cmd/at_cmd.c

```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
-					       payload_len);
-				} else {
-					LOG_ERR("Response buffer not large "
-						"enough");
+		payload_len = get_return_code(buf, &ret);
 
-					ret.code  = -EMSGSIZE;
-				}
-
-				response_buf_len = 0;
-				response_buf     = NULL;
+		if (ret.state == AT_CMD_NOTIFICATION) {
+			ret.callback = notification_handler;
+		}
 
+		/* Allocate a buffer if we need to, verify its size otherwise */
+		if (ret.state == AT_CMD_NOTIFICATION ||
+		    (current_cmd.buf == NULL &&
+		     current_cmd.flags & AT_CMD_BUF_RESP)){
+			ret.buf = k_malloc(payload_len);
+			if (ret.buf == NULL) {
+				LOG_ERR("Failed to allocate response buffer");
+				ret.code  = -ENOMEM;
 				goto next;
 			}
+			ret.flags |= AT_RESP_ALLOC;
+		} else if (current_cmd.buf != NULL) {
+			if (current_cmd.buf_len < payload_len) {
+				LOG_ERR("Response buffer not large enough");
+				ret.code  = -EMSGSIZE;
+				goto next;
+			}
+			ret.buf = current_cmd.buf;
 		}
 
-		if (payload_len == 0) {
-			goto next;
-		}
-
-		if (ret.state == AT_CMD_NOTIFICATION) {
-			item->callback = notification_handler;
-		} else {
-			item->callback = current_cmd_handler;
+		/* Fill the buffer if there is one to fill at this point*/
+		if (ret.buf != NULL) {
+			memcpy(ret.buf, buf, payload_len);
 		}
 next:
-		/* If no callback was set, free the item.
-		 * Otherwise, work queue callback will free it.
-		 */
-		if (item->callback == NULL) {
-			k_mem_slab_free(&rsp_work_items, (void **)&item);
-		} else {
-			k_work_init(&item->work, callback_worker);
-			k_work_submit(&item->work);
-		}
-
-		/* Notify back only if command was sent. */
-		if ((k_sem_count_get(&cmd_pending) == 0) &&
-		    (ret.state != AT_CMD_NOTIFICATION)) {
-			current_cmd_handler = NULL;
+		distribute_resp(&current_cmd, &ret);
 
-			k_msgq_put(&return_code_msq, &ret, K_FOREVER);
+		/* We have now handled a command if it was not a notification */
+		if (ret.state != AT_CMD_NOTIFICATION) {
+			complete_cmd();
 		}
 	}
 }
 
-static inline int at_write(const char *const cmd, enum at_cmd_state *state)
+/* Allocate and copy cmd to a new buffer */
+static char *buffer_cmd(const char *cmd)
 {
-	int bytes_sent;
-	int bytes_to_send = strlen(cmd);
-	struct return_state_object ret;
-
-	LOG_DBG("Sending command %s", log_strdup(cmd));
-
-	bytes_sent = send(common_socket_fd, cmd, bytes_to_send, 0);
-
-	if (bytes_sent == -1) {
-		LOG_ERR("Failed to send AT command (err:%d)", errno);
-		ret.code  = -errno;
-		ret.state = AT_CMD_ERROR;
-	} else {
-		LOG_DBG("Awaiting response for %s", log_strdup(cmd));
-		k_msgq_get(&return_code_msq, &ret, K_FOREVER);
-		LOG_DBG("Bytes sent: %d", bytes_sent);
-
-		if (bytes_sent != bytes_to_send) {
-			LOG_ERR("Bytes sent (%d) was not the "
-				"same as expected (%d)",
-				bytes_sent, bytes_to_send);
-		}
-	}
+	char *temp;
 
-	if (state) {
-		*state = ret.state;
+	temp = k_malloc(strlen(cmd) + 1);
+	if (temp == NULL) {
+		return NULL;
 	}
-
-	return ret.code;
+	memcpy(temp, cmd, strlen(cmd) + 1);
```

Comment:
We deal with a string, you can use `strcpy`.

Reply:
Done